runtime/io.h

```diff
@@ -133,43 +133,71 @@ void IONAME(SetSign)(Cookie, const char *, std::size_t);
 // Any item can be transferred by means of a descriptor; unformatted
 // transfers to/from contiguous blocks can avoid the descriptor; and there
 // are specializations for the common scalar types.
-// Functions with Boolean results return false when the I/O statement
-// has encountered an error or end-of-file/record condition; the others
-// will return a zero value.
+// These functions return false when the I/O statement has encountered an error
+// or end-of-file/record condition.  Once the statement has encountered
+// an error, following items will be ignored; but compiled code should
+// check for errors and avoid the following items when a READ statement
+// might have data dependences between items, as in
+//   READ(*,ERR=666) N, (A(J),J=1,N)
+// or when input/output items might raise a spurious fault, as in
+//   WRITE(*,ERR=666) X, CRASHIFCALLEDANDERRORPRESENT()
 bool IONAME(OutputDescriptor)(Cookie, const Descriptor &);
 bool IONAME(InputDescriptor)(Cookie, const Descriptor &);
 bool IONAME(OutputUnformattedBlock)(Cookie, const char *, std::size_t);
 bool IONAME(InputUnformattedBlock)(Cookie, char *, std::size_t);
 bool IONAME(OutputInteger64)(Cookie, std::int64_t);
-std::int64_t IONAME(InputInteger64)(Cookie, int kind = 8);
+bool IONAME(InputInteger64)(Cookie, std::int64_t &, int kind = 8);
 bool IONAME(OutputReal32)(Cookie, float);
-float IONAME(InputReal32)(Cookie);
+bool IONAME(InputReal32)(Cookie, float &);
 bool IONAME(OutputReal64)(Cookie, double);
-double IONAME(InputReal64)(Cookie);
+bool IONAME(InputReal64)(Cookie, double &);
 bool IONAME(OutputComplex32)(Cookie, float, float);
 bool IONAME(OutputComplex64)(Cookie, double, double);
 bool IONAME(OutputAscii)(Cookie, const char *, std::size_t);
 bool IONAME(InputAscii)(Cookie, char *, std::size_t);
 bool IONAME(OutputLogical)(Cookie, bool);
 bool IONAME(InputLogical)(Cookie);
 
-// Result extraction; these can be called at any time during an
-// I/O data transfer statement to check for errors, or after all
-// of the data transfers are complete to acquire the final status.
-void IONAME(GetIoMsg)(Cookie, char *, std::size_t);  // IOMSG=
-void IONAME(GetStatus)(Cookie, char *, std::size_t);  // STATUS=
-int IONAME(GetIostat)(Cookie);  // IOSTAT=
 std::size_t IONAME(GetSize)(Cookie);  // SIZE=
-bool IONAME(IsEnd)(Cookie);
-bool IONAME(IsEor)(Cookie);
-bool IONAME(IsErr)(Cookie);
-
-// The cookie value must not be used after calling this function.
-// If an error has occurred and not been noticed by an inquiry
-// function like GetIOSTAT() or IsERR(), this function will
-// terminate the image with a message.
+
+// GetIoMsg() does not modify its argument unless an error or
+// end-of-record/file condition is present.
+void IONAME(GetIoMsg)(Cookie, char *, std::size_t);  // IOMSG=
+
+// One of these two functions must be called to end an I/O data
```

Comment:
[Possible recommendation at the end of this comment.]

If one of these calls must possibly call TerminateImage, is there much benefit in defining two calls?  Or to ask this a different way, there are 15 nonempty subsets of {Iostat, Err, End, Eor} specifiers.  For 6 of these subsets, the call to EndIoStatementWithErr must be followed by a conditional call to TerminateImage in case the appropriate specifier is absent.

Does TerminateImage need access to the Cookie to determine termination message content?

If the library knew from the beginning which of IOSTAT, ERR, END, and EOF were present, statement end processing would be simplified.  In particular, image termination could be done immediately with no further interaction with compiled code.  But yes, this does have a cost that the design seems to be trying hard to avoid, such as adding a common argument and source coordinate information to probably all statement begin calls, or defining a boilerplate call that must be called for all statements.

As a variation of all of these, how about defining an {IOSTAT, ERR, END, EOF} argument on a single common termination call.  That avoids adding this info to just one call rather than all begin calls, but gives the library the information it needs to choose termination or not while it still has information available for messaging.  The arguments could be made optional so that they need not be supplied for the 9 cases where it is known that the library will not be required to terminate the image.

Reply:
Ok, I will try to do that.

Reply:
Please see the latest commit.

Reply:
After more consideration, I've moved the error handling mask and source location arguments up into all of the `Begin...()` calls.  We want the library to be able to terminate the image as soon as an unhandled error/end condition arises rather than risking unrelated and confusing crashes that might arise from attempts to address and transfer further items, I think.  Please take another look.